src/modules/auth/token.service.spec.ts

```diff
@@ -0,0 +1,84 @@
+import { Controller, Get, INestApplication, Injectable, InternalServerErrorException, MiddlewareConsumer, Module, RequestMethod, UnauthorizedException } from '@nestjs/common';
+import { Test, TestingModule } from '@nestjs/testing';
+import { Request } from 'express';
+import { NextFunction } from 'express-serve-static-core';
+import request from 'supertest';
+import * as jwt from 'jsonwebtoken';
+import { TokenPayload } from './token.service';
+
+// Mock token service class
+@Injectable()
+class TokenService {
+
+    handleOriginCheck = async (req: Request, res: Response, next: NextFunction) => {
+        let token = null;
+        if (req.headers['authorization']) {
+            token = req.headers['authorization'].replace('Bearer ', '');
+        } else {
+            token = req?.cookies?.token
+        }
+
+        if (token) {
+            const decodedToken = jwt.decode(token) as TokenPayload;
+            const isBrowserRequestFromAuthenticatedOrigin = decodedToken.origin === req.headers['origin']
+            const isServerRequestOrGETFromSameDomain = req.headers['origin'] === undefined
+            if (isBrowserRequestFromAuthenticatedOrigin || isServerRequestOrGETFromSameDomain) {
+                next()
+            } else {
+                throw new InternalServerErrorException('Something went wrong')
+            }
+        } else {
+            throw new UnauthorizedException('Unauthorized')
+        }
+    }
+}
+
+// mock controller
+@Controller()
+class TestController {
+    @Get('test')
+    test() {
+        return "RETURN_VALUE";
+    }
+}
+
+// mock the middleware
+@Module({ providers: [TokenService], exports: [TokenService] })
+class TestModule {
+    constructor(private readonly tokenService: TokenService) { }
+
+    configure(consumer: MiddlewareConsumer) {
+        consumer
+            .apply((req, res, next) => this.tokenService.handleOriginCheck(req, res, next))
+            .exclude(
+                { path: '/login', method: RequestMethod.ALL },
+                { path: '/refresh_token', method: RequestMethod.ALL }
+            )
+            .forRoutes({ path: '/*', method: RequestMethod.ALL });
+    }
+};
+
+describe('TokenService', () => {
+    let app: INestApplication;
+
+    beforeEach(async () => {
+        const module: TestingModule = await Test.createTestingModule({
+            imports: [TestModule],
+            controllers: [TestController],
+        }).compile();
+
+        app = module.createNestApplication();
+        await app.init();
+    });
+
+    it('should have same GET request ORIGIN header as stored in JWT token ', async () => {
```

Comment:
Is the `origin` header set for this test? I think it is probably not being set as it is a node request? If so, I think this test name is misleading

Reply:
for the current version, the origin header is undefined which also matches  our check, will add an origin with headers so naming should be consistent then.

Reply:
As a stretch goal, it would be nice to have tests of other HTTP methods (e.g. `POST`)

Reply:
updated by latest [commit](https://github.com/energywebfoundation/iam-cache-server/pull/145/commits/34318563453d02a5a399f48163b79890cb244c15)